### DISCUSSION TITLE
Unable to visit dynamic route on Next.js only in production

### DISCUSSION THREAD

**Discussion by diegoleonardoro** [vercel/next.js#50514] [Help] [answered]
### Summary

I am on a Next.js project and my team and I are having an issue when I try to dynamically render a page. This only happens on production when I deploy the project on the Google Cloud Platform.

My folder structure is the following:

      pages
         users
            [userId].js
         index.js
The issue happens when I try to visit the [userId].js route from index.js

Both pages are making use of getServerSideProps()

index.js:
      
      export const getServerSideProps =  async(context)=>{ 
        const client = buildClient(context);
        const { data } = await client.get("/api/residents");
        return { props:{residents: data} };
      }
      
      const LandingPage = ({ currentUser, users }) => {
      
        const { addProduct, cartItems, increase } = useContext(CartContext);
      
        const usersList = users.map((user) => {
      
          return (
            <tr key={resident.id}>
              <td className={styles.tableCell}>{user.user.name}</td>
              <td className={styles.tableCell}>{user.neighborhood}</td>
              <td className={styles.tableCell}>
                {user.neighborhoodDescription}
              </td>
              <td className={styles.tableCell}>
                <Link href= "/users/[userId]" as={`/users/${user.id}`} > { // <<-----
                  View
                </Link>
              </td>
            </tr>
          );
        });
        return (
           ....
        );
      };
      
[userId].js
        
        export async function getServerSideProps(context ) {
          // Fetch necessary data for the blog post using params.id
          const {residentId} = context.query;
          const userData = await getResidentData(residentId, context);
        
          return {
            props: {
              resident:userData.data || null,
              products: []
            },
          };
        }
      
      const ResidentShow = ({currentUser, resident, products}) => {       
          …
      }
why am I not able to render [userId].js when I try to visit that page from index.js. This only happens when I deploy on the Google Cloud. Locally, the project runs smoothly.

**Comment by devjiwonchoi** [comment]
I don't see where you are handling the `currentUser` props at `[userId].js`. Could you elaborate for me?

```
export async function getServerSideProps(context ) {
  // Fetch necessary data for the blog post using params.id
  const {residentId} = context.query;
  const userData = await getResidentData(residentId, context);

  return {
    props: {
      resident:userData.data || null,
      products: []
      currentUser:    // maybe missing here?
    },
  };
}

const ResidentShow = ({currentUser, resident, products}) => {       
  …
}
```

**Reply by diegoleonardoro** [reply]
**** Thanks for your answer. I am not making use of currentUser inside of [userId].js. 

currentUser is being passed from the AppComponent, but I do not know if my implementation is correct:


        export const getServerSideProps = async (context) => {
          const client = buildClient(context);
          const { data } = await client.get("/api/users/currentuser");
        
          let pageProps = {};
        
          if (context.Component.getServerSideProps) {
            pageProps = await context.Component.getServerSideProps(context, client, data.currentUser);
          }
        
          return {
            props: {
              ...data,
              pageProps,
            },
          };
        }
        
        const AppComponent = ({ Component, pageProps, currentUser }) => {
        
          const [showForm, setShowForm] = useState(false);
        
          const blurVal = showForm ? 12 : 0;
        
          const [blur, setBlur] = useState("0");
        
          function changeBlur(value) {
            setBlur(value);
          }
        
          return (
            <ProductsContextProvider>
              <CartContextProvider>
                <SelectedProductContextProvider>
                  <div
                    style={{
                      position: "relative",
                      filter: `blur(${blur}px)`,
                      WebkitFilter: `blur(${blur}px)`, // For Safari
                      backgroundColor: "rgba(255, 255, 255, 0.9)", // Transparent white
                      zIndex: "9999",
                      fontSize:"12px"
                    }}
                  >
                    <div className={styles.wrapper}>
                      <div
                        style={{ filter: `blur(${blurVal}px)` }}
                        className={styles.wrapperDiv}
                      >
                        <Header
                          currentUser={currentUser}
                          setShowFormValue={setShowForm}
                          showFormValue={showForm}
                        />
                        <div className={styles.container}>
                          <Component currentUser={currentUser} {...pageProps} />
                        </div>
                      </div>
                      <Form
                        display={showForm}
                        setShowFormValue={setShowForm}
                        currentUser={currentUser}
                      ></Form>
                    </div>
                  </div>
                  <SingleProduct changeBlur={changeBlur}> </SingleProduct>
                </SelectedProductContextProvider>
              </CartContextProvider>
            </ProductsContextProvider>
          );
        
        };

**Reply by devjiwonchoi** [reply]
Assuming that your API Endpoint and everything inside the `getServerSideProps` is flawless,
does your
```
if (context.Component.getServerSideProps) {
  pageProps = await context.Component.getServerSideProps(context, client, data.currentUser);
}
```
returns a valid output on production?

I would recommend you wrap the `fetch` section around with a `try catch` statement to verify that the `props` is passed correctly to the `AppComponent`.

**Reply by devjiwonchoi** [reply, marked as answer]
## Static Site Generation

Although Dynamic Routes are available by using `getServerSideProps` only, I would recommend you to use `getStaticPaths` and `getStaticProps` for [Static Site Generation](https://nextjs.org/docs/pages/building-your-application/rendering/static-site-generation) since it can improve your build time and accessibility of the users.

Example:

```js
// pages/repo/[name].tsx

import type {
  InferGetStaticPropsType,
  GetStaticProps,
  GetStaticPaths,
} from 'next';
 
type Repo = {
  name: string;
  stargazers_count: number;
};
 
export const getStaticPaths: GetStaticPaths = async () => {
  return {
    paths: [
      {
        params: {
          name: 'next.js',
        },
      }, // See the "paths" section below
    ],
    fallback: true, // false or "blocking"
  };
};
 
export const getStaticProps: GetStaticProps<{
  repo: Repo;
}> = async () => {
  const res = await fetch('https://api.github.com/repos/vercel/next.js');
  const repo = await res.json();
  return { props: { repo } };
};
 
export default function Page({
  repo,
}: InferGetStaticPropsType<typeof getStaticProps>) {
  return repo.stargazers_count;
}
```

## Incremental Static Regeneration

Another option I would recommend is to try [Incremental Static Regeneration](https://nextjs.org/docs/pages/building-your-application/data-fetching/incremental-static-regeneration) since it provides the benefit of static generation without rebuilding the entire site, and the scaling of numerous pages of Dynamic Route.

> Incremental Static Regeneration (ISR) enables you to use static-generation on a per-page basis, without needing to rebuild the entire site. With ISR, you can retain the benefits of static while scaling to millions of pages.

## Reference
- [Page Router Static Site Generation](https://nextjs.org/docs/pages/building-your-application/rendering/static-site-generation)
- [Incremental Static Regeneration](https://nextjs.org/docs/pages/building-your-application/data-fetching/incremental-static-regeneration)
